specs/Menu.md

```diff
@@ -1,731 +1,813 @@
 
-The menu should provide some help for orienting its contents either vertically, like a left nav, or horizontally, like a nav bar. Ideally this will just be a prop that gets put into context so each item can decide how it should appear. It's possible that the menu should enforce direction itself.
+const menu = (
+  <Menu trigger={trigger}>
+    <MenuItem icon={<FileIcon />}>Option 1</MenuItem>
+    <MenuItem icon={<BellIcon />}>Option 2</MenuItem>
+    <MenuItem icon={<LinkIcon />}>Option 3</MenuItem>
+  <Menu>
+)
+```
 
-##### Focus
+```html
+<!-- expected DOM output  -->
+<!-- TODO positioning -->
+<button aria-haspopup="true" aria-expanded="true" id="trigger">Open menu</button>
+<div role="menu" aria-labelledby="trigger">
+  <div role="menuitem" tabindex="0">
+    <span role="presentation"><svg>FileIcon</svg></span>
+    Option 1
+  </div>
+  <div role="menuitem" tabindex="0">
+    <span role="presentation"><svg>BellIcon</svg></span>
+    Option 2
+  </div>
+  <div role="menuitem" tabindex="0">
+    <span role="presentation"><svg>LinkIcon</svg></span>
+    Option 3
+  </div>
+</div>
+```
 
-By default Focus should go to the tab stops and it's up to each menuItem to decide if it has tabindex=0. An example of a menuItem that shouldn't be focusable would be a Divider.
+### Sections
+
+```typescript
+const trigger = <button> Open menu </button>
+
+const menu = (
+  <Menu trigger={trigger}>
+    <MenuItem>Option 1</MenuItem>
+    <MenuSection title="Section title">
+      <MenuItem>Section Option 1</MenuItem>
+      <MenuItem>Section Option 2</MenuItem>
+      <MenuItem>Section Option 3</MenuItem>
+    <MenuSection />
+  <Menu>
+)
+```
 
-However for the menu we should provide a focus wrapper that is able to track focus using role="menuitem" to determine what item should get focused.
+```html
+<!-- expected DOM output  -->
+<!-- TODO positioning -->
+<button aria-haspopup="true" aria-expanded="true" id="trigger">Open menu</button>
+<div role="menu" aria-labelledby="trigger">
+  <div role="menuitem" tabindex="0">Option 1</div>
+  <div role="separator"></div>
+  <div role="presentation" aria-hidden="true" id="sectionid">Section title</div>
+  <div role="group" aria-labelledby="sectionid">
+    <div role="menuitem" tabindex="-1">Section Option 1</div>
+    <div role="menuitem" tabindex="-1">Section Option 2</div>
+    <div role="menuitem" tabindex="-1">Section Option 3</div>
+  </div>
+  <div role="separator"></div>
+</div>
+```
 
-The menu does not need to provide a native way to be controlled without getting focus directly. Like above, there should be a simple utility wrapper that could be written to implement this.
+### Submenus
+
+```typescript
+const trigger = <button> Open menu </button>
+
+const menu = (
+  <Menu trigger={trigger}>
+    <MenuItem>Option 1</MenuItem>
+    <SubMenu content="Open submenu">
+      <MenuItem>Option 1</MenuItem>
+      <MenuItem>Option 2</MenuItem>
+      <MenuItem>Option 3</MenuItem>
+    <SubMenu>
```

Comment:
It'd be great to have the API for SubMenu be parallel to the API for MenuItem and Menu (with regards to content and triggering), rather than introducing the `content` slot for SubMenu only, and not for MenuItem.

If we went with the current `trigger` proposal, it would look something like:
```tsx
<Menu trigger={<Button>Open menu</Button>}>
  <MenuItem>Option 1</MenuItem>
  <SubMenu trigger={<MenuItem>Open submenu</MenuItem>}>
    <MenuItem>Option 1</MenuItem>
    <MenuItem>Option 2</MenuItem>
    <MenuItem>Option 3</MenuItem>
  </SubMenu>
</Menu>
```

Another option might be to have subMenu be a slot on MenuItem, and the content of the MenuItem would still be its child:
```tsx
<Menu trigger={<Button>Open menu</Button>}>
  <MenuItem>Option 1</MenuItem>
  <MenuItem subMenu={
      <SubMenu>
        <MenuItem>Option 1</MenuItem>
        <MenuItem>Option 2</MenuItem>
        <MenuItem>Option 3</MenuItem>
      </SubMenu>}>
    Open submenu
  </SubMenu>
</Menu>
```

Reply:
@behowell I recommend you come to the trigger deepdive !

I would avoid putting `MenuItems` in a slot. I would prefer no slots tbh. It's a 'sacrifice' I made in this API proposal to have some way of associating an extraneous element to the declaration of the actual menu as children. I would like a API structure to reflect DOM structure as much as possible.]

Several ppl have approached me to help me improve and they will be attending the deepdive. You should definitely come and weigh in too :-)